src/examples/learn_quant/monotonicity.py

```diff
@@ -0,0 +1,254 @@
+
+    if np.all(quantifier) or not np.any(quantifier):
+        return 1
+    p_q_true = sum(quantifier) / len(quantifier)
+    p_q_false = 1 - p_q_true
+    q_ent = -p_q_true * np.log2(p_q_true) - p_q_false * np.log2(p_q_false)
+
+    def get_preds(num_arr, num):
+        """Given an array of ints, and an int, get all predecessors of the
+        model corresponding to int.
+        Returns an array of same shape as num_arr, but with bools
+        """
+        return num_arr[num, :]
+
+    def has_true_pred(num_arr, y):
+        return np.any(y * num_arr)
+
+    # where necessary?
+    true_preds = np.where(np.dot(submembership, quantifier) >= 1, 1, 0)
+
+    p_pred = sum(true_preds) / len(true_preds)
+    p_nopred = 1 - p_pred
+
+    # TODO: make this elegant! solve nan problems
+    q_pred = sum(quantifier.T * true_preds) / len(quantifier)
+    q_nopred = sum(quantifier.T * (1.0 - true_preds)) / len(quantifier)
+    noq_pred = sum((1.0 - quantifier.T) * true_preds) / len(quantifier)
+    noq_nopred = sum((1.0 - quantifier.T) * (1.0 - true_preds)) / len(quantifier)
+
+    pred_logs = np.log2([noq_pred, q_pred] / p_pred)
+    pred_logs[pred_logs == -np.inf] = 0
+    nopred_logs = np.log2([noq_nopred, q_nopred] / p_nopred)
+    nopred_logs[nopred_logs == -np.inf] = 0
+    ent_pred = -np.nansum(np.array([noq_pred, q_pred]) * pred_logs)
+    ent_nopred = -np.nansum(np.array([noq_nopred, q_nopred]) * nopred_logs)
+    cond_ent = ent_pred + ent_nopred
+
+    # return 0 if q_ent == 0 else 1 - (cond_ent / q_ent)
+    return (1.0 - cond_ent / q_ent)[0, 0]
+
+
+def calculate_monotonicity(universe, expressions, down=False):
+
+    metrics = {}
+    concept_lattice = calculate_lattice(universe)
+
+    membership = sparse.lil_matrix((len(universe), len(expressions)), dtype=int)
+    submembership = sparse.lil_matrix((len(universe), len(universe)), dtype=int)
+    model_dictionary = {}
+    for model_id, model in enumerate(universe.referents):
+        model_dictionary[model.name] = model_id
+    for model_id, model in enumerate(universe.referents):
+        for expression_id, quantifier_expression in enumerate(expressions):
+            metrics[str(quantifier_expression)] = {}
+            if model in quantifier_expression.meaning.referents:
+                membership[model_id, expression_id] = 1
+        sub_ids = list(
+            map(
+                model_dictionary.__getitem__,
+                get_sub_structures(concept_lattice, [model.name]),
+            )
+        )
+        for sub_id in sub_ids:
+            submembership[sub_id, model_id] = 1
+
+    membership = membership.todense()
+    submembership = submembership.todense()
+
+    for expression_id, quantifier_expression in enumerate(expressions):
+        print("Calculating monotonicity for: ", quantifier_expression)
+        metrics[str(quantifier_expression)]["monotonicity"] = (
+            upward_monotonicity_entropy(submembership, membership[:, expression_id])
```

Comment:
why is submembership the first argument here?

Reply:
As opposed to the second you mean? I can switch them?

Reply:
In the original method `upward_monotonicity_entropy` that this is modified from, the first argument was `all_models`, which was an array containing all the quantifiermodels in the universe, not just the submodels.  But maybe some other logic has changed there?

Reply:
Ah I see. I might have been testing something and the code drifted into the test. I'll check it out